streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -253,6 +259,86 @@ public VersionedRecord<byte[]> get(final Bytes key, final long asOfTimestamp) {
         return null;
     }
 
+    // Visible for testing
+    ValueIterator<VersionedRecord<byte[]>> get(final Bytes key, final long fromTimestamp, final long toTimestamp, final boolean isAscending) {
+
+        Objects.requireNonNull(key, "key cannot be null");
+        validateStoreOpen();
+
+        final List<VersionedRecord<byte[]>> queryResults = new ArrayList<>();
```

Comment:
> Not sure if we should materialize the result in-memory -- it there is a long history, it might put a lot of memory pressure on the JVM. -- Instead, we could return an Iterator of "segment iterators" (only for segments that contain data, or might contain data) and to the iteration "lazy / on-demand" instead.
> 
> It's an somewhat open question for a single-key query, but we don't want to materialize the result for range-key queries for sure...
> 
> (Cf. range queries over windowed stores for which we do the same -- at least high level -- given how the segment store work, we might not want to use an actual RocksDBIterator -- cf my top level review comment)

I won't change the code but will add the topic to the KIP. Thanks a lot.

Reply:
Seems are still materializing the result in-memory? Given the new snapshot implementation, I think we can avoid this, by pushing most of the logic to do actual `get()` into latest/segment stores into the iterator itself, to execute gets lazily, too.